### DISCUSSION TITLE
Type narrowing based on a key of a TypedDict tuple element

### DISCUSSION THREAD

**Discussion by InSyncWithFoo** [microsoft/pyright#8071] [General]
Consider the following example (playground links: [Mypy](https://mypy-play.net/?flags=strict&mypy=1.10.0&python=3.12&gist=1893fd9a55d89e7944f23f9c3098ffe6), [Pyright](https://pyright-play.net/?pyrightVersion=1.1.365&pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgCYAiSAxjADRQAySMxIAhgDYCwAUL1RzYBnIVACCARgAURBGUo0AlAC5eUdbBLLGzVpwDaAcgmGAur36CR4gEwySFajBVqN8OdqYt2HIzbMWPALCogBC0rLyTi48GlAARhLaQjAggcHWoXaRjkqqsRrxNsmpgQDCkPGobDBIYCgSUAC8sACuCBzE%2BpL04eY8FRBVKDV1KDbNbR1dYja9Nv2BpMTAUMBSKTXE2oPDo-WNAD5Qu9W19TaKUAC0AHxQAHL1267qbPTxk5ssgXFIq98ugAGUxGdzEMzNFrGQz5OJxEDEABuxE4AH1wRsYFsrlAAMRQCCIODw0lNe4wdqdboSXoSfqkxlMxkEhBwEBINAAC3wjPJU2pPSgfSgx0p026c2FC1e8MRKPRmLYuLiBKJbKgAGooGyOdz8PzJKLbLKEcjURwMSQpPEVRo1cStTr2Zyec17uFjVlfhp-lA2GCSJCmtCTHDGfKLVa5FiceoHRrtbrXQaKVSZrThfTjeLBVKsgyI%2BbFdblaSEySky79e7xBJTRpIyWY7by4THVW9W7%2BZ7jlkgA)):

```python
from typing import TypedDict, Literal

class A1(TypedDict):
    type: Literal['1']

class A2(TypedDict):
    type: Literal['2']

class B1(TypedDict):
    b1: str

class B2(TypedDict):
    b2: str

Combination1 = tuple[A1, B1]
Combination2 = tuple[A2, B2]

# https://stackoverflow.com/q/78577720
def f(state: Combination1 | Combination2) -> None:
    a, b = state

    if state[0]['type'] == '1':
        reveal_type(state)  # mypy           => tuple[A1, B1]
                            # pyright        => tuple[A1, B1] | tuple[A2, B2]
        reveal_type(a)      # mypy + pyright => A1 | A2
        reveal_type(b)      # mypy + pyright => B1 | B2

    if a['type'] == '1':
        reveal_type(state)  # mypy + pyright => tuple[A1, B1] | tuple[A2, B2]
        reveal_type(a)      # mypy + pyright => A1
        reveal_type(b)      # mypy + pyright => B1 | B2
```

I'd argue Mypy's behaviour is more useful. Is this a bug or an intentional decision that Pyright chose not to narrow `state` to `Combination1`?

**Comment by erictraut** [comment]
This isn't a supported type guard pattern. For a complete list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

The type guard pattern `state[0]['type'] == '1'` narrows the expression `state[0]`. It doesn't narrow the expression `state`. That would be theoretically possible, but it would require some additional special-case logic. I'd consider adding this support if there is signal from other pyright users that it's something they want to see. What you're doing here strikes me as quite unusual and probably doesn't apply to many other users.
